docs/features/nullable-reference-types.md

```diff
@@ -0,0 +1,119 @@
+Nullable Reference Types
+=========
+References to reference types may be nullable, non-nullable, or null-oblivious. (The document uses the shorthand `?`, `!`, and `~`.)
+
+## Annotations
+In source, nullable reference types are annotated with `?`.
+```c#
+string? OptString; // may be null
+List<object?>? OptListOptElements; // list may be null, elements may be null
+```
+In metadata, nullable reference types are annotated with a `[Nullable]` attribute.
+```c#
+namespace System.Runtime.CompilerServices
+{
+    public sealed class NullableAttribute : Attribute
+    {
+        public NullableAttribute() { }
+        public NullableAttribute(bool[] b) { }
+    }
+}
+```
+If annotations are enabled in a module, a `[module: Nullable]` attribute is included.
+```c#
+// C# representation of metadata
+[module: Nullable]
+[Nullable] string OptString; // string?
+[Nullable(new[] { true, true })] List<object> OptListOptElements; // List<object?>?
+```
+`NullableAttribute` cannot be used explicitly in source so the type declaration can be synthesized by the compiler if it is not included in the compilation.
+
+Unannotated reference types are non-nullable or null-oblivious depending whether the containing scope includes `[NonNullTypes]`.
+```c#
+namespace System.Runtime.CompilerServices
+{
+    public sealed class NonNullTypesAttribute : Attribute
+    {
+        public NonNullTypesAttribute(bool enabled = true) { }
+    }
+}
+```
+If there is no `[NonNullTypes]` attribute in any containing scope, including the module, reference types are non-nullable.
+```c#
+[NonNullTypes(false), Nullable(new[] { false, true })] string[] Oblivious; // string?[]~
+[NonNullTypes(true), Nullable(new[] { false, true })] string[] NotNull; // string?[]!
+```
+`NonNullTypesAttribute` can be referenced in source so the type declaration must be provided explicitly to the compilation and should be defined in the framework.
+
+## Declaration warnings
+_Describe warnings reported for declarations in initial binding._
+
+## Flow analysis
+Flow analysis is used to infer the nullability of variables within executable code. The inferred nullability of a variable is independent of it's declared nullability.
+
+### Warnings
+_Describe set of warnings. Differentiate W warnings._
+
+### Null tests
+_Describe the set of tests that affect flow state._
+
+### Conversions
+_Describe valid top-level and variance conversions._
+_Describe warnings from user-defined conversions._
+
+### Assignment
+Nullability of the righthand side flows to the lefthand side unchanged. Warnings are reported for top-level and nested nullability mismatches.
```

Comment:
Sometimes, the warning is a `W` warning. Either add the details, or a remark to fill-in more details.

Reply:
"mismatch": what are we comparing?
The left-hand side could be a local/parameter/API, in which case the declared type is what matters (not the state).

Nit: I'm not sure the spelling but "righthand" is not recognized by my local dictionary, whereas "right-hand side" is.

Reply:
Can you take a note to follow-up on parameters and APIs? It's not just locals that produce W warnings.

---
In reply to: [194867721](https://github.com/dotnet/roslyn/pull/27721#discussion_r194867721) [](ancestors = 194867721)